### ISSUE TITLE
ComboBox - add ValueMatchBinding property

### ISSUE THREAD

**Issue by tomasherceg** [riganti/dotvvm#156]
The viewmodel looks like this:

```
public List<Country> Countries { get; set; }
public Country SelectedCountry { get; set; }
```

The view looks like this:

```
<dot:ComboBox DataSource="{value: Countries}" SelectedValue="{value: SelectedCountry}" />
```

However, the binding doesn't work really well because when deserializing the JSON from the server, the deserializer doesn't know that the SelectedCountry and e.g. Countries[2] are the same instances.
Therefore we need a property to detect, which instances are equal. 

The usage would be like this:

```
<dot:ComboBox DataSource="{value: Countries}" SelectedValue="{value: SelectedCountry}"
    ValueMatchBinding="{value: Id}" />
```

**Comment by martinsura** [comment]
Btw is planned use ValueMatchBinding="{value:Id}" for mulit-select ComboBox?

**Comment by tomasherceg** [comment]
We were also thinking about the following names:

* `KeyBinding`
* `ItemKeyBinding`
* `ItemCompareBinding`
* `EqualityBinding`

**Comment by tomasherceg** [comment]
To be consistent with the Business Pack naming, we will use `ItemKeyBinding`.

**Comment by exyi** [comment]
There are actually more problems than just matching values by id when you have the same instance multiple times in the same view model. For example, the client-side deserialization changes both the selected object and the object in the collection even though you have changed only one on the server. And it does much crazier things if you change both of them... And it gets even worse when you have the collection bound to the ComboBox, although I'm not sure exactly why.

I think that we basically can't fix this issue, without changing the serialization/deserialization process significantly (to something like [lave](https://github.com/jed/lave)), and the only thing we can do is to warn users in the docs. Maybe we could throw an exception or write a warning to console when one object is used multiple times in the view, but I don't know how to efficiently detect it in JS (no reference-based hashtable)

**Comment by djanosik** [comment]
Well, we must make sure we don't use the same instance on multiple places in view model. That's what we do in BP and the `ItemKeyBinding` works quite well. Of course, you shouldn't have multiple objects with the same key.

**Comment by exyi** [comment]
@djanosik But the knockout puts the original object from the collection to selected value by reference, which causes that the object is multiple times in the view model. Do you deep copy all the things in BussinessPack?

**Comment by djanosik** [comment]
Exactly, we make deep copies of objects added to selected values. Because its very similar to what happens when you set SelectedValue in postback. It's deserialized as another instance.

**Comment by tomasherceg** [comment]
Solved in DotVVM 2.0 - `ItemKeyBinding`.
